lib/puppetserver/ca/generate_action.rb

```diff
@@ -0,0 +1,249 @@
+require 'optparse'
+require 'openssl'
+require 'puppetserver/utils/file_utils'
+require 'puppetserver/settings/ttl_setting'
+
+module Puppetserver
+  module Ca
+    class GenerateAction
+      CA_EXTENSIONS = [
+        ["basicConstraints", "CA:TRUE", true],
+        ["keyUsage", "keyCertSign, cRLSign", true],
+        ["subjectKeyIdentifier", "hash", false],
+        ["authorityKeyIdentifier", "keyid:always", false]
+      ]
+      DEFAULT_SIGNING_DIGEST = OpenSSL::Digest::SHA256.new
+
+      SUMMARY = "Generate a root and intermediate signing CA for Puppet Server"
+      BANNER = <<-BANNER
+Usage:
+  puppetserver ca generate [--help|--version]
+  puppetserver ca generate [--config PATH]
+
+Description:
+Generate a root and intermediate signing CA for Puppet Server
+and store generated CA keys, certs, and crls on disk.
+
+To determine the target location, the default puppet.conf
+is consulted for custom values. If using a custom puppet.conf
+provide it with the --config flag
+Options:
+BANNER
+
+      def initialize(logger)
+        @logger = logger
+      end
+
+      def run(input)
+        # Validate config_path provided
+        config_path = input['config']
+        errors = Puppetserver::Utils::FileUtilities.validate_file_paths([config_path])
+        log_errors_and_return(errors) if errors.any?
+
+        # Load, resolve, and validate puppet config settings
+        puppet = PuppetConfig.parse(config_path)
+        log_errors_and_return(puppet.errors) if puppet.errors.any?
+
+        # Generate root and intermediate ca and put all the certificates, crls,
+        # and keys where they should go.
+        generate_root_and_intermediate_ca(puppet)
```

Comment:
from what I can tell the `generate_root_and_intermediate_ca` method doesn't need the actual puppet object just the settings from it. It would probably be cleaner to pass `puppet.settings` into this method and then have something like:
```rb
  def generate_root_and_intermediate_ca(settings)
    private_key_length = settings[:keylength]
    # ...
```

Reply:
Yup.  Totally didnt catch that.  Thanks!